packages/ui/src/Modal.tsx

```diff
@@ -0,0 +1,166 @@
+ */
+type ModalActionDisabledProps =
+  | {
+      actionDisabledTooltip: ButtonDisabledProps['disabledTooltip']
+      actionDisabled: ButtonDisabledProps['disabled']
+    }
+  | {
+      actionDisabledTooltip?: never
+      actionDisabled?: never
+    }
+
+type ModalActionIconProps =
+  | {
+      actionIconLeft: ButtonAccessibleIconLeftProps['iconLeft']
+      actionIconLeftAriaLabel: ButtonAccessibleIconLeftProps['iconLeftAriaLabel']
+    }
+  | {
+      actionIconLeft?: never
+      actionIconLeftAriaLabel?: never
+    }
+
+type ModalActionCoreProps =
+  | {
+      // Note: Source of `onClick` is nullable property from ButtonHTMLAttributes<HTMLButtonElement>
+      onAction: NonNullable<ButtonProps['onClick']>
+      actionLabel: string
+    }
+  | { onAction?: never; actionLabel?: never }
+
+export type ModalActionProps = ModalActionCoreProps & {
+  actionKind?: ButtonProps['kind']
+} & ModalActionDisabledProps &
+  ModalActionIconProps
+
+export type ModalProps = {
+  closable?: boolean
+  children?: ReactNode
+  title: string
+  onClose?: ReactModalProps['onRequestClose']
+  footerClassName?: string
+} & DataTestProp &
+  ModalActionProps &
+  Exclude<ReactModalProps, 'onRequestClose' | 'shouldFocusAfterRender' | 'shouldReturnFocusAfterClose'>
+
+/*
+ * ### Purpose
+ * Occasionally there's a user-story that's not a part of the main user flow. Such action can be contained in a Modal.
+ * Modals can contain components like forms, menus, tables etc.
+ * Usually modals are used to drive a complex action. In case there is a simple action (e.g. confirmation), consider using a Dialog.
+ *
+ * ### General Info
+ * - Modal always contains a title, icon "X" in the header, content and active buttons in the footer.
+ * - No interactions on the underlying page can be performed while a Modal.
+ * - Content beneath the modal is covered by a "blanket".
+ * - To close the Modal, use "Cancel" button in the footer, "X" button in the header, press "Esc" key or click anywhere on the "blanket".
+ */
+export const Modal: FC<ModalProps> = ({
+  'data-test': dataTest,
+  children,
+  className,
+  footerClassName,
+  closable = true,
+  onClose,
+  overlayClassName,
+  title,
+  ...restWithActionButton
+}) => {
+  const {
+    actionKind,
+    actionLabel,
+    // Disabled
+    actionDisabled = false,
+    actionDisabledTooltip,
+    // Icon
+    actionIconLeft,
+    actionIconLeftAriaLabel,
+    // Rest
+    onAction,
+    ...rest
+  } = restWithActionButton
+
+  return (
+    <ReactModal
+      contentLabel={title}
+      className={cn(styles.modal, className)}
+      overlayClassName={cn(styles.overlay, overlayClassName)}
+      data-test={dataTest}
+      onRequestClose={onClose}
+      shouldCloseOnEsc={closable}
+      shouldCloseOnOverlayClick={closable}
+      shouldFocusAfterRender
+      shouldReturnFocusAfterClose
+      {...rest}
+    >
+      <div className={styles.outline} />
```

Comment:
Why do we need it? AFAICS, the modal itself is not focusable. Only the elements inside.

Reply:
It actually is focusable. `react-modal` exposes a prop `shouldFocusAfterRender` to control the behaviour.

This just mirrors out current MC behavior.

Reply:
@jankoritak AFAIK, according to ARIA guidelines the focus should be on the first focusable element (https://www.w3.org/TR/wai-aria-practices-1.1/#keyboard-interaction-7), not on the modal itself. I suggest we add `focusTarget: Ref<HTMLELement>` prop to the `Modal` and focus it if provided.
WDYT?

Reply:
@jankoritak could you point me to any guidelines suggesting to focus the modal itself?

Reply:
I'm not aware of any. I'm not opposing this idea, I'm rather on board with it. 👍 The reason why I included the property was to mirror original MC behaviour.

>  I suggest we add focusTarget: Ref<HTMLELement> prop to the Modal and focus it if provided.

We can go this path. But I'm wondering whether arbitrary `autoFocus` on given content element is not enough for the job here. WDYT?

Reply:
Yep. `autofocus` totally works :)

Reply:
I'm thinking, what should happen when the modal has no focusable element? That will lead to certain confusion of the user.

Actually, with our Modal/Dialog, that will never happen, as there is always at least `Cancel` button. But according to following source, I think we should implement following.

Source:
- https://www.w3.org/TR/wai-aria-practices-1.1/#keyboard-interaction-7
> If a dialog contains the final step in a process that is not easily reversible, such as deleting data or completing a financial transaction, it may be advisable to set focus on the least destructive action, especially if undoing the action is difficult or impossible. The Alert Dialog Pattern is often employed in such circumstances.

>If a dialog is limited to interactions that either provide additional information or continue processing, it may be advisable to set focus to the element that is likely to be most frequently used, such as an OK or Continue button.

#### Modal
- In case there are no `actions` provided, we `autoFocus` the `Cancel` button.
- In case there are actions provided and none of them has `autoFocus`, keep the `autoFocus` on `Cancel` button.

#### Dialog
- Note: Built on top of Modal
- In case the action is `primary`, `autoFocus` the action
- In case the action is `danger`, `autoFocus` `Cancel` button

WDYT?

Reply:
@jankoritak sounds great! Let's do exactly that! I would also suggest adding `autofocus: false` prop to the modal itself to allow developers set it manually to some item in content if necessary. WDYT?

Reply:
Oh man, that autofocus sounds really great!

Reply:
Makes sense. Addressed in https://github.com/hazelcast/frontend-shared/pull/148/commits/139abb63216a52d3673a10fc8f99be8adc0a5e0c.